In [124]:
import pandas as pd
import time

In [53]:
ingredients_df = pd.read_pickle('data/ingredients.pkl')
effects_df = pd.read_pickle('data/effects.pkl')

In [3]:
ingredients[1:50]

name        id   expansion  \
1        Ancestor Moth Wing  xx0059ba   Dawnguard   
2         Ash Creep Cluster  xx01cd74  Dragonborn   
3          Ash Hopper Jelly  xx01cd71  Dragonborn   
4           Ashen Grass Pod  xx016e26  Dragonborn   
5                Bear Claws  0006bc02        Base   
6                       Bee  000a9195        Base   
7              Beehive Husk  000a9191        Base   
8            Bleeding Crown  0004da20        Base   
9               Blisterwort  0004da25        Base   
10      Blue Butterfly Wing  000727de        Base   
11            Blue Dartwing  000e4f0c        Base   
12     Blue Mountain Flower  00077e1c        Base   
13                Boar Tusk  xx01cd6f  Dragonborn   
14                Bone Meal  00034cdd        Base   
15              Briar Heart  0003ad61        Base   
16      Burnt Spriggan Wood  xx01cd6e  Dragonborn   
17           Butterfly Wing  000727e0        Base   
18               Canis Root  0006abcb        Base   
19     Charred Skeever Hide  00052695        Base   
20             Chaurus Eggs  0003ad56        Base   
21  Chaurus Hunter Antennae  xx0183b7   Dawnguard   
22            Chicken's Egg  00023d77        Base   
23            Creep Cluster  000b2183        Base   
24         Crimson Nirnroot  000b701a        Base   
25      Cyrodilic Spadetail  00106e19        Base   
26             Daedra Heart  0003ad5b        Base   
27                Deathbell  000516c8        Base   
28          Dragon's Tongue  000889a2        Base   
29              Dwarven Oil  000f11c0        Base   
30                Ectoplasm  0003ad63        Base   
31                Elves Ear  00034d31        Base   
32     Emperor Parasol Moss  xx01ff75  Dragonborn   
33         Eye of Sabre Cat  0006bc07        Base   
34               Falmer Ear  0003ad5d        Base   
35    Felsaad Tern Feathers  xx03cd8e  Dragonborn   
36               Fire Salts  0003ad5e        Base   
37              Fly Amanita  0004da00        Base   
38            Frost Mirriam  00034d32        Base   
39              Frost Salts  0003ad5f        Base   
40                   Garlic  00034d22        Base   
41             Giant Lichen  0007e8c1        Base   
42              Giant's Toe  0003ad64        Base   
43             Gleamblossom  xx00b097   Dawnguard   
44                Glow Dust  0003ad73        Base   
45         Glowing Mushroom  0007ee01        Base   
46                Grass Pod  00083e64        Base   
47            Hagraven Claw  0006b689        Base   
48        Hagraven Feathers  0003ad66        Base   
49             Hanging Moss  00057f91        Base   

                                           how_to_get  \
1               Collected by catching Ancestor Moths.   
2   Harvested from creep cluster plants unique to ...   
3                    Collected from dead ash hoppers.   
4   Harvested from spiky grass plants unique to So...   
5         Collected from various types of dead Bears.   
6       Collected from beehives and by catching Bees.   
7                            Collected from beehives.   
8   Harvested from Bleeding Crown, a mushroom foun...   
9   Harvested from Blisterwort mushrooms found in ...   
10  Collected by catching Blue Butterflies. Often ...   
11  Collected by catching Dragonflies found near r...   
12  Harvested from the blue variety of Mountain Fl...   
13    Collected from dead bristlebacks and rieklings.   
14  Collected from undead such as skeletons, and D...   
15               Collected from Forsworn Briarhearts.   
16               Collected from dead burnt spriggans.   
17  Collected from catching Monarch Butterflies. O...   
18  Harvested from Canis Root, found in The Rift a...   
19  Food, collected from dead Skeevers found on ro...   
20  Collected from Chaurus Egg Sacs, found in Falm...   
21  Collected from dead Chaurus Hunters and Chauru...   
22  Food, harvested from Chicken nests in various ...   
23  Harvested from Creep Cluster, found in the vol...   
24  

In [4]:
effects

name        id  \
1               Cure Disease  000AE722   
2                Cure Poison  00109ADD   
3              Damage Health  0003EB42   
4             Damage Magicka  0003A2B6   
5       Damage Magicka Regen  00073F2B   
6             Damage Stamina  0003A2C6   
7       Damage Stamina Regen  00073F2C   
8                       Fear  00073F20   
9         Fortify Alteration  0003EB24   
10            Fortify Barter  0003EB23   
11             Fortify Block  0003EB1C   
12      Fortify Carry Weight  0003EB01   
13       Fortify Conjuration  0003EB25   
14       Fortify Destruction  0003EB26   
15        Fortify Enchanting  0003EB29   
16            Fortify Health  0003EAF3   
17       Fortify Heavy Armor  0003EB1E   
18          Fortify Illusion  0003EB27   
19       Fortify Light Armor  0003EB1F   
20       Fortify Lockpicking  0003EB21   
21           Fortify Magicka  0003EAF8   
22          Fortify Marksman  0003EB1B   
23        Fortify One-handed  0003EB19   
24        Fortify Pickpocket  0003EB20   
25       Fortify Restoration  0003EB28   
26          Fortify Smithing  0003EB1D   
27             Fortify Sneak  0003EB22   
28           Fortify Stamina  0003EAF9   
29        Fortify Two-handed  0003EB1A   
30                    Frenzy  00073F29   
31              Invisibility  0003EB3D   
32                     Light  FExxx846   
33   Lingering Damage Health  0010AA4A   
34  Lingering Damage Magicka  0010DE5F   
35  Lingering Damage Stamina  0010DE5E   
36                 Night Eye  FExxx803   
37                 Paralysis  00073F30   
38             Ravage Health  00073F26   
39            Ravage Magicka  00073F27   
40            Ravage Stamina  00073F23   
41         Regenerate Health  0003EB06   
42        Regenerate Magicka  0003EB07   
43        Regenerate Stamina  0003EB08   
44               Resist Fire  0003EAEA   
45              Resist Frost  0003EAEB   
46              Resist Magic  00039E51   
47             Resist Poison  00090041   
48              Resist Shock  0003EAEC   
49            Restore Health  0003EB15   
50           Restore Magicka  0003EB17   
51           Restore Stamina  0003EB16   
52                      Slow  00073F25   
53          Spell Absorption  FExxx812   
54            Waterbreathing  0003AC2D   
55          Weakness to Fire  00073F2D   
56         Weakness to Frost  00073F2E   
57         Weakness to Magic  00073F51   
58        Weakness to Poison  00090042   
59         Weakness to Shock  00073F2F   

                                          description  \
1                                 Cures all diseases.   
2                  Stops poison's continuing effects.   
3               Causes <mag> points of poison damage.   
4        Drains the target's Magicka by <mag> points.   
5   Decrease the target's Magicka regeneration by ...   
6         Drain the target's Stamina by <mag> points.   
7   Decrease the target's Stamina regeneration by ...   
8   Creatures and people up to level <mag> flee fr...   
9   Alteration spells last <mag>% longer for <dur>...   
10  You haggle for <mag>% better prices for <dur> ...   
11  Blocking absorbs <mag>% more damage for <dur> ...   
12  Carrying capacity increases by <mag> for <dur>...   
13  Conjurations spells last <mag>% longer for <du...   
14  Destruction spells are <mag>% stronger for <du...   
15  For <dur> seconds, items are enchanted <mag>% ...   
16  Health is increased by <mag> points for <dur> ...   
17  Increase Heavy Armor skill by <mag> points for...   
18  Illusion spells are <mag>% stronger for <dur> ...   
19  Increases Light Armor skill by <mag> points fo...   
20    Lockpicking is <mag>% easier for <dur> seconds.   
21  Magicka is increased by <mag> points for <dur>...   
22      Bows do <mag>% more damage for <dur> seconds.   
23  One-handed weapons do <mag>% more damage for <...   
24  Pickpocketing is <mag>% easier for <dur> seconds.   
25  Restoration spells are <mag>% stronger for <du...   
26  For <dur> seconds

$$x_1 - 2x_2 + 3x_3 + y_1 + y_2$$

$$x_1 + 2x_2 \le 5$$
$$x_1 + x_3 \le 2.5$$
$$2 \le x_1 + y_1 + 2y_2 \le 4.2$$
$$2 \le x_3 + y_1 + 2y_2 \le 3$$
$$ y_1, y_2 \ge 0$$

In [6]:
import numpy as np
from cylp.cy import CyCbcModel, CyClpSimplex
from cylp.py.modeling.CyLPModel import CyLPModel, CyLPArray

model = CyLPModel() # using the clp.py.modeling.CyLPModel modeling tool to create the LP

# add variables
x = model.addVariable('x', 3, isInt=True)
y = model.addVariable('y', 2)

A = np.matrix([[1., 2., 0],[1., 0, 1.]])
B = np.matrix([[1., 0, 0], [0, 0, 1.]])
D = np.matrix([[1., 2.],[0, 1]])
a = CyLPArray([5, 2.5])
b = CyLPArray([4.2, 3])
x_u= CyLPArray([2., 3.5])

model += A*x <= a
model += 2 <= B * x + D * y <= b
model += y >= 0
model += 1.1 <= x[1:3] <= x_u

c = CyLPArray([1., -2., 3.])

model.objective = c * x + 2 * y.sum()

s = CyClpSimplex(model) # Create the cylp.cy.CySimplex object, using the created model as input

cbcModel = s.getCbcModel() # Create the Cbc model, which is capable of solving MIPs using B&B

cbcModel.solve()
print(cbcModel.status)
sol_x = cbcModel.primalVariableSolution['x']
print(sol_x)
(abs(sol_x - np.array([0, 2, 2])) <= 10**-6).all()


solution
[0. 2. 2.]


/home/eric/anaconda3/lib/python3.7/site-packages/scipy/sparse/lil.py:512: FutureWarning: future versions will not create a writeable array from broadcast_array. Set the writable flag explicitly to avoid this warning.
  if not i.flags.writeable or i.dtype not in (np.int32, np.int64):
/home/eric/anaconda3/lib/python3.7/site-packages/scipy/sparse/lil.py:518: FutureWarning: future versions will not create a writeable array from broadcast_array. Set the writable flag explicitly to avoid this warning.
  if not x.flags.writeable:


True

There are 109 ingredients. All potions/poisons are made by combining 2 or 3 unique ingredients. Thus, there are ${ 109 \choose 2 } = 5886$ two-ingredient combinations and ${109 \choose 3} = 209934$ three-ingredient combinations.

In [23]:
import itertools
comb = list(itertools.combinations(ingredients.name, 4)) 
len(comb)

5563251

We now create a data frame for all two-ingredient potions:

In [67]:
def get_effects(ingredients):
    effects_list = []
    for ingredient in ingredients:
        effects = list(ingredients_df.loc[ingredients_df.name == ingredient, 
                                  ['effect_1', 'effect_2', 'effect_3', 'effect_4']].values[0])
        effects_list.append(effects)
        
    possible_effects = {}
    for effects in effects_list:
        for effect in effects:
            possible_effects[effect] = possible_effects.get(effect, 0) + 1
    
    created_effects = []
    for effect in possible_effects.keys():
        if possible_effects[effect] >= 2:
            created_effects.append(effect)
    
    return created_effects

In [129]:
tic = time.perf_counter()
potion_df = pd.DataFrame()
potion_data = []
for i in range(len(ingredients_df)):
    toc = time.perf_counter()
    first = ingredients_df.name[i]
    print(f'Current First Ingredient: {first} \t Total elapsed time: {toc - tic:.2f} seconds')
    for j in range(i+1, len(ingredients_df)):
        second = ingredients_df.name[j]
        effects = get_effects([first, second])
        potion_data.append([first, second, 'NA', effects])
        for k in range(j+1, len(ingredients_df)):
            third = ingredients_df.name[k]
            effects = get_effects([first, second, third])
            potion_data.append([first, second, third, effects])
toc = time.perf_counter()
print(f'Total elapsed time: {toc - tic:.2f} seconds')

Current First Ingredient: Abecean Longfin 	 Total elapsed time: 0.00 seconds
Current First Ingredient: Ancestor Moth Wing 	 Total elapsed time: 13.06 seconds
Current First Ingredient: Ash Creep Cluster 	 Total elapsed time: 25.82 seconds
Current First Ingredient: Ash Hopper Jelly 	 Total elapsed time: 38.40 seconds
Current First Ingredient: Ashen Grass Pod 	 Total elapsed time: 50.83 seconds
Current First Ingredient: Bear Claws 	 Total elapsed time: 62.91 seconds
Current First Ingredient: Bee 	 Total elapsed time: 74.85 seconds
Current First Ingredient: Beehive Husk 	 Total elapsed time: 86.52 seconds
Current First Ingredient: Bleeding Crown 	 Total elapsed time: 98.22 seconds
Current First Ingredient: Blisterwort 	 Total elapsed time: 109.44 seconds
Current First Ingredient: Blue Butterfly Wing 	 Total elapsed time: 120.42 seconds
Current First Ingredient: Blue Dartwing 	 Total elapsed time: 131.19 seconds
Current First Ingredient: Blue Mountain Flower 	 Total elapsed time: 141.75 sec

In [130]:
potion_df = potion_df.append(potion_data, ignore_index = True)

In [192]:
potion_df.columns = ['first_ingredient', 'second_ingredient', 'third_ingredient', 'effects']
potion_df.head(20)

first_ingredient   second_ingredient      third_ingredient  \
0   Abecean Longfin  Ancestor Moth Wing                    NA   
1   Abecean Longfin  Ancestor Moth Wing     Ash Creep Cluster   
2   Abecean Longfin  Ancestor Moth Wing      Ash Hopper Jelly   
3   Abecean Longfin  Ancestor Moth Wing       Ashen Grass Pod   
4   Abecean Longfin  Ancestor Moth Wing            Bear Claws   
5   Abecean Longfin  Ancestor Moth Wing                   Bee   
6   Abecean Longfin  Ancestor Moth Wing          Beehive Husk   
7   Abecean Longfin  Ancestor Moth Wing        Bleeding Crown   
8   Abecean Longfin  Ancestor Moth Wing           Blisterwort   
9   Abecean Longfin  Ancestor Moth Wing   Blue Butterfly Wing   
10  Abecean Longfin  Ancestor Moth Wing         Blue Dartwing   
11  Abecean Longfin  Ancestor Moth Wing  Blue Mountain Flower   
12  Abecean Longfin  Ancestor Moth Wing             Boar Tusk   
13  Abecean Longfin  Ancestor Moth Wing             Bone Meal   
14  Abecean Longfin  Ancestor Moth Wing           Briar Heart   
15  Abecean Longfin  Ancestor Moth Wing   Burnt Spriggan Wood   
16  Abecean Longfin  Ancestor Moth Wing        Butterfly Wing   
17  Abecean Longfin  Ancestor Moth Wing            Canis Root   
18  Abecean Longfin  Ancestor Moth Wing  Charred Skeever Hide   
19  Abecean Longfin  Ancestor Moth Wing          Chaurus Eggs   

                                              effects  
0                                                  []  
1                                    [Damage Stamina]  
2                                 [Weakness to Frost]  
3                                     [Fortify Sneak]  
4                              [Damage Magicka Regen]  
5                                                  []  
6                                     [Fortify Sneak]  
7                                [Weakness to Poison]  
8                                    [Damage Stamina]  
9   [Damage Stamina, Fortify Conjuration, Damage M...  
10                                                 []  
11        [Fortify Conjuration, Damage Magicka Regen]  
12                                                 []  
13              [Damage Stamina, Fortify Conjuration]  
14                                                 []  
15                             [Damage Magicka Regen]  
16                                                 []  
17                                   [Damage Stamina]  
18                                                 []  
19                               [Weakness to Poison]

In [172]:
effect_counts = [len(effects) for effects in potion_df['effects']]
sum([effect_count == 5 for effect_count in effect_counts]) / len(effect_counts)
quintuple_indices = [i for i in range(len(effect_counts)) if effect_counts[i] == 5]

Of the 215820 possible combinations, 43.55% of them have no effect, 40.24% have a single effect, 13.85% have a double effect, 1.99% have a triple effect, 0.36% have a quadruple effect, and 0.00003% (just 5) have a quintuple effect.

In [202]:
valid_potion_indices = [i for i in range(len(potion_df)) if potion_df.effects[i] != []]
valid_potions = potion_df.loc[valid_potion_indices]
valid_potions.reset_index(inplace=True)
valid_potions.to_pickle('data/potions.pkl')

In [203]:
valid_potions

index first_ingredient       second_ingredient  \
0            1  Abecean Longfin      Ancestor Moth Wing   
1            2  Abecean Longfin      Ancestor Moth Wing   
2            3  Abecean Longfin      Ancestor Moth Wing   
3            4  Abecean Longfin      Ancestor Moth Wing   
4            6  Abecean Longfin      Ancestor Moth Wing   
...        ...              ...                     ...   
121821  215807       Void Salts          Wisp Wrappings   
121822  215808       Void Salts          Wisp Wrappings   
121823  215812            Wheat               White Cap   
121824  215814            Wheat          Wisp Wrappings   
121825  215815            Wheat  Yellow Mountain Flower   

              third_ingredient                                 effects  
0            Ash Creep Cluster                        [Damage Stamina]  
1             Ash Hopper Jelly                     [Weakness to Frost]  
2              Ashen Grass Pod                         [Fortify Sneak]  
3                   Bear Claws                  [Damage Magicka Regen]  
4                 Beehive Husk                         [Fortify Sneak]  
...                        ...                                     ...  
121821                      NA                          [Resist Magic]  
121822  Yellow Mountain Flower                          [Resist Magic]  
121823  Yellow Mountain Flower  [Fortify Health, Damage Stamina Regen]  
121824  Yellow Mountain Flower  [Fortify Health, Damage Stamina Regen]  
121825                      NA  [Fortify Health, Damage Stamina Regen]  

[121826 rows x 5 columns]

The model is the following:

#### Parameters

$P = \{1, \dots, 121826\}$: Set of potions

$I = \{1, \dots, 109\}$: Set of ingredients

$I_p \subset{I}$: Set of ingredients required to make one unit of potion $p$ (note that $|I_p| \leq 3 \,\,\forall p \in P$)

$z_i \in \mathbb{Z}^+$: Amount of ingredient $i$ available

$c_p \in \mathbb{R}^+$: gold cost of potion $p$

#### Decision Variables

$y_p \in \mathbb{Z}^+$: Number of potions $p$ created

$x_{ip} \in \mathbb{Z}^+$: Amount of ingredient $i$ used to create potion $p$

#### Model

$$ \max \sum_{i \in p} c_p y_p$$
such that:
$$y_p \leq x_{ip} \qquad \forall i \in I_p,  \forall p \in P$$
$$\sum_{\{p \mid i \in I_p\}} x_{ip} \leq z_i \qquad \forall i \in I$$


An additional sophistication for the model is to allow the purchase of common ingredients, which would amount to adding a integer decision variable for each such ingredient, adding them to the RHSs of the second constraints, and subtracting the costs in the objective.

In [205]:
potions = pd.read_pickle('data/potions.pkl')

In [214]:
potions[potions['first_ingredient'] == "Chicken's Egg"][potions['second_ingredient'] == "Nightshade"]

/home/eric/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


index first_ingredient second_ingredient        third_ingredient  \
59610  109005    Chicken's Egg        Nightshade                      NA   
59611  109006    Chicken's Egg        Nightshade                Nirnroot   
59612  109007    Chicken's Egg        Nightshade         Nordic Barnacle   
59613  109008    Chicken's Egg        Nightshade         Orange Dartwing   
59614  109009    Chicken's Egg        Nightshade                   Pearl   
59615  109010    Chicken's Egg        Nightshade         Pine Thrush Egg   
59616  109011    Chicken's Egg        Nightshade            Poison Bloom   
59617  109012    Chicken's Egg        Nightshade   Powdered Mammoth Tusk   
59618  109013    Chicken's Egg        Nightshade  Purple Mountain Flower   
59619  109014    Chicken's Egg        Nightshade     Red Mountain Flower   
59620  109015    Chicken's Egg        Nightshade             River Betty   
59621  109016    Chicken's Egg        Nightshade        Rock Warbler Egg   
59622  109017    Chicken's Egg        Nightshade         Sabre Cat Tooth   
59623  109018    Chicken's Egg        Nightshade              Salmon Roe   
59624  109019    Chicken's Egg        Nightshade               Salt Pile   
59625  109020    Chicken's Egg        Nightshade          Scaly Pholiota   
59626  109021    Chicken's Egg        Nightshade              Scathecraw   
59627  109022    Chicken's Egg        Nightshade        Silverside Perch   
59628  109023    Chicken's Egg        Nightshade            Skeever Tail   
59629  109024    Chicken's Egg        Nightshade       Slaughterfish Egg   
59630  109025    Chicken's Egg        Nightshade    Slaughterfish Scales   
59631  109026    Chicken's Egg        Nightshade           Small Antlers   
59632  109027    Chicken's Egg        Nightshade             Small Pearl   
59633  109028    Chicken's Egg        Nightshade             Snowberries   
59634  109029    Chicken's Egg        Nightshade               Spawn Ash   
59635  109030    Chicken's Egg        Nightshade              Spider Egg   
59636  109031    Chicken's Egg        Nightshade            Spriggan Sap   
59637  109032    Chicken's Egg        Nightshade        Swamp Fungal Pod   
59638  109033    Chicken's Egg        Nightshade                 Taproot   
59639  109034    Chicken's Egg        Nightshade          Thistle Branch   
59640  109035    Chicken's Egg        Nightshade         Torchbug Thorax   
59641  109036    Chicken's Egg        Nightshade              Trama Root   
59642  109037    Chicken's Egg        Nightshade               Troll Fat   
59643  109038    Chicken's Egg        Nightshade           Tundra Cotton   
59644  109039    Chicken's Egg        Nightshade            Vampire Dust   
59645  109040    Chicken's Egg        Nightshade              Void Salts   
59646  109041    Chicken's Egg        Nightshade                   Wheat   
59647  109042    Chicken's Egg        Nightshade               White Cap   
59648  109043    Chicken's Egg        Nightshade          Wisp Wrappings   
59649  109044    Chicken's Egg        Nightshade  Yellow Mountain Flower   

                                                 effects  
59610   [Damage Magicka Regen, Lingering Damage Stamina]  
59611  [Resist Magic, Damage Magicka Regen, Lingering...  
59612  [Damage Magicka Regen, Waterbreathing, Lingeri...  
59613   [Damage Magicka Regen, Lingering Damage Stamina]  
59614   [Damage Magicka Regen, Lingering Damage Stamina]  
59615   [Damage Magicka Regen, Lingering Damage Stamina]  
59616   [Damage Magicka Regen, Lingering Damage Stamina]  
59617   [Damage Magicka Regen, Lingering Damage Stamina]  
59618   [Damage Magicka Regen, Lingering Damage Stamina]  
59619  [Damage Magicka Regen, Lingering Damage Stamin...  
59620   [Damage Magicka Regen, Lingering Damage Stamina]  
59621   [Damage Magicka Regen, Lingering Damage Stamina]  
59622   [Damage Magicka Regen, Lingering Damage Stamina]  
59623   [Damage Magicka Regen, Lingering Damage Stamina]  
59624 